atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/KvsSweepProgressQueueTest.java

```diff
@@ -0,0 +1,122 @@
+
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.impl.InMemoryKeyValueService;
+
+public class KvsSweepProgressQueueTest {
+    private KeyValueService kvs;
+    private KvsSweepQueueProgress progress;
+
+    @Before
+    public void setup() {
+        kvs = new InMemoryKeyValueService(true);
+        progress = new KvsSweepQueueProgress(kvs);
+    }
+
+    @Test
+    public void canReadInitialNumberOfShards() {
+        assertThat(progress.numberOfShards()).isEqualTo(1);
+    }
+
+    @Test
+    public void canUpdateNumberOfShards() {
+        progress.numberOfShards();
+        progress.updateNumberOfShards(128);
+        assertThat(progress.numberOfShards()).isEqualTo(128);
+    }
+
+    @Test
+    public void attemptingToDecreaseNumberOfShardsIsNoop() {
+        progress.numberOfShards();
+        progress.updateNumberOfShards(64);
+        progress.updateNumberOfShards(32);
+        assertThat(progress.numberOfShards()).isEqualTo(64);
+    }
+
+    @Test
+    public void increasingNumberOfShardsAbove128Throws() {
+        progress.numberOfShards();
+        assertThatThrownBy(() -> progress.updateNumberOfShards(256)).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void canReadInitialSweptTimestamp() {
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(0L);
+    }
+
+    @Test
+    public void canUpdateSweptTimestamp() {
+        progress.lastSweptTimestamp(10, true);
+        progress.updateLastSweptTimestamp(10, true, 1024L);
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(1024L);
+    }
+
+    @Test
+    public void attemptingToDecreaseSweptTimestampIsNoop() {
+        progress.lastSweptTimestamp(10, true);
+        progress.updateLastSweptTimestamp(10, true, 1024L);
+        progress.updateLastSweptTimestamp(10, true, 512L);
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(1024L);
+    }
+
+    @Test
+    public void updatingTimestampForOneShardDoesNotAffectOthers() {
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(0L);
+        assertThat(progress.lastSweptTimestamp(20, true)).isEqualTo(0L);
+        progress.updateLastSweptTimestamp(10, true, 1024L);
+        assertThat(progress.lastSweptTimestamp(20, true)).isEqualTo(0L);
+
+        progress.updateLastSweptTimestamp(20, true, 512L);
+        assertThat(progress.lastSweptTimestamp(20, true)).isEqualTo(512L);
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(1024L);
+    }
+
+    @Test
+    public void updatingTimestampForOneConsistencyDoesNotAffectOther() {
+        assertThat(progress.lastSweptTimestamp(10, true)).isEqualTo(0L);
+        assertThat(progress.lastSweptTimestamp(10, false)).isEqualTo(0L);
+        progress.updateLastSweptTimestamp(10, true, 128L);
+        assertThat(progress.lastSweptTimestamp(10, false)).isEqualTo(0L);
+
+        progress.updateLastSweptTimestamp(10, false, 32L);
```

Comment:
Does it make sense to have different conservative and thorough timestamps per shard?

Reply:
Depends on the implementation - but we're already treating them as separate queues (so we have 2 queues per shard) based on the timestamps writer in part 2, so I think it makes sense.